### ISSUE TITLE
Manifest "book.json" Hpub support

### ISSUE THREAD

**Issue by folletto** [Simbul/baker#102]
As an author I want to create a manifest to define the metadata of the book.
As a user I want to be able to know exactly what book is it.

The manifest file is a JSON file included in the book package, named "book.json" and containing the book definition itself.

We defined in addition to the "standard" feature a way to specify device properties, something inherited by various standards extensions specification: any top level property starting with "x-" defines an extension. In this case: "x-baker".

This is the structure at the current draft:

```
{
  "hpub": 1,
  "title": "The Study in Scarlet",
  "author": ["Arthur Conan Doyle"],
  "creator": ["D. Casali", "M. Colombo", "A. Morandi", "F. Fortes"],
  "date: "2011-08-23",
  "url": "book://bakerframework.com/books/arthurconandoyle-thestudyinscarlet",

  "orientation": "both",
  "zoomable": false,

  "-baker-background": "#000000",
  "-baker-vertical-bounce": true,
  "-baker-index-bounce": true,
  "-baker-index-height": 150,
  "-baker-media-autoplay": true,

  "-treesaver-loopArticles": "true",

  "contents": [
    "Book Cover.html",
    "Book Index.html",
    {
        "url": "Front-Article.html",
        "title": "The Wonders of Hpub",
        "author": "The community"
    },
    "Part1-01.html",
    "Part1-02.html",
    "Part1-03.html",
    "Part1-04.html",
    "Part1-05.html",
    "Part1-06.html",
    "Part1-07.html",
    "Part2-01.html",
    "Part2-02.html",
    "Part2-03.html",
    "Part2-04.html",
    "Part2-05.html",
    "Part2-06.html",
    "Part2-07.html",
    "Colophon.html",
  ]
}
```

**Comment by fortes** [comment]
We'd love to adopt the same JSON format for Treesaver, as I think there are many benefits to our two projects using the exact same format. However, we have a few requirements that would require some changes to this format in order to meet our needs. I think they're fairly compatible with what you have here.

## Support Objects in the `pages` array, as well as strings

Developers can add per-article metadata within Treesaver, and we support this via JSON within our current TOC ([example](https://github.com/Treesaver/treesaver/blob/master/test/functional/simple/toc.json)). Having this information exist solely in the file (say via `<meta>`) isn't an option since we cannot fetch every file in the array for performance reasons.

My proposal is to modify the `pages` array you have here to support objects as well as strings, meaning the following example would be valid:

``` js
"pages": [
  {
    "url": "Part1.html",
    "title": "Part One",
    "author": "John Example",
    "my_custom_field": "Custom value"
  },
  "Part2.html"
]
```

From a code standpoint, it is a matter of checking the type of each entry in the array.

Related: Treesaver supports sectioning/nested articles. This is something we do via a property within the article object, which can be ignored by clients that do not support it.

## Support publication-wide custom metadata

Much as we allow authors to provide custom values per article, we'd also like to enable doing this across a publication. I suggest allowing these at the top level of the object, as peers to `title`, etc. So for example:

``` js
{
  "title": "My Title",
  "author": "John Example",
  "creator": "...",
  // etc

  "my_custom_field": "Custom value",
  "my_custom_field2": ["value_1", "value_2"],

  // Properties, pages, etc
}
```

Note that this data is _not_ Treesaver-specific, rather it is metadata about the publication itself.

Alternatively, these custom fields could live within a separate object, but our preference is to have them top-level for simplicity. This also make extensibility and future proofing a bit easier, and follows the conventions of `package.json`, which allow for arbitrary field addition as well.

## Renaming the `pages` array

This is a minor point, but I believe the term `pages` is a bit overloaded. When creating a paginated layout, a single "web page" can be split into several "pages" (screens?), and the terms can get a little confusing (e.g. page 3 of `foo.html`). I'm wondering if a different name could be used, perhaps `contents`, `parts`, `toc`, or `articles`?

## `properties` vs. `x-baker`, `x-treesaver`, etc

I'm not quite sure I understand the separation between the `properties` object and user-agent specific settings. The example above has `pinchToZoom` and `orientation` within the `properties` object, while `background` and `verticalBounce` exist within the `x-baker` object. It seems like `orientation`, etc are fairly tablet-specific, while `background` seems quite general (`indexHeight` seems perfect for `x-baker` though).

Borrowing inspiration from CSS, what if we merged into a `settings` object, and prefixed properties which are user-agent specific:

``` js
"settings": {
  "background": "#000000",
  "-baker-indexHeight": 150,
  // Etc
}
```

I really like the work you've done so far, and I hope we can get to a common format here. Thanks!

**Comment by folletto** [comment]
Hi Fortes :)

First of all, thanks for the suggestions! These are very constructive contribution. :)

Now, point by point:
1. **Objects in pages**: it adds a bit of complexity but allows also the simpler form, and I think it's a good compromise. I'd just add that we should formalize a bit which fields are required in the hash form. Given what you wrote, I think that "url", "title" and "author" are the fields that we could add in the HPub standard and leave the rest to extensions. OK.
2. **Custom metadata**: here I'm not sure, because it means that from opening a HPub you wouldn't know what's in the standard - and so, must be supported - and what's not. I was thinking along the lines of the CSS specification, that so far I think it's working pretty well, adding prefixes for extensions, like "x-". My doubt is that if we say in the standard "and then you can use whatever field you want with any name... well... it's not a standard anymore, no? :) Do you have a strong position in favor of allowing custom metadata in that way and not inside a x-name block, specified by each reader technology?
3. **Rename pages array**: I see your point, and between the various choices I think that "content" is the best solution, because it's precise enough, but doesn't suggest a specific form of the output. OK.
4. **Properties vs x-baker**: This part was a big discussion we had during a few of our calls. The idea is this: "properties" defines things that are related to the book, in a generic way. So orientation and pinchToZoom are in this perspective properties of the book and the rendering engine should read these and adapt accordingly. Instead, the extensions are add-ons, out of the standard, so a book might not use them at all. While not specifying the orientation is a problem, because all the tablets out there can manage that, ignoring "background" is just something that affect the specific "Baker" rendering engine, and another reader, for example Treesaver, can safely ignore them. In a similar fashion if Treesaver wants to support certain extended parameters, it can safely read the x-treesaver hash and get them, while Baker will read the book still very well by ignoring them. As for Baker, we are trying to get rid of those, that are mostly related to technical limitations, so they will just disappear as soon as possible. The CSS suggestion you are making is basically the same, we just group extensions in specific hashes (think also that some x-baker or x-treesaver features might not be "properties" or "settings" at all, maybe they are metadata or content or protections or else). Now that I explained the difference, do you still think we should merge everything in a single "settings" hash? If yes, could you list he pros of that choice? :)

**Comment by fortes** [comment]
_Moving these out of order for better flow:_

**Renaming Array**

I have a preference for `contents` over `content` since it is an array with multiple items, and `content` gives a slight feeling of being the text content of the book.

**Objects in Pages**

Glad you're OK with this, since it is definitely key for our scenarios. As for the required fields within, I consider `url` the only required field (maps neatly to using just a string). The other fields can be recommended and widely-recognized, but otherwise it's up to the author to determine what can go in there (this dovetails into the next point):

**Re: Custom Metadata**

The metadata in this case isn't user-agent specific, but rather it is metadata for the _content_. The thinking for this custom metadata (be it at the root level or within an individual article) is that it is extra information provided by the author of the content / end developer -- _not_ by tools or by reader software. So imagine someone wishes to add a `license` property, or perhaps they wish to add a `permalink` property, which points to a web-accessible URL for an article (or even a `shortlink`). 

There are a few parallels here in terms of other formats -- I mentioned [package.json](http://wiki.commonjs.org/wiki/Packages/1.0), which has the notion of required and optional fields, along with the ability to add custom fields to the object as well. This works pretty well for ad-hoc extensibility and some level of futureproofing (new fields that are added later do not make the manifest file invalid for older versions).

Another example is with HTML -- new tags are simply ignored if they are unknown. `<meta>` and `<link>` tags work in a similar way in terms of being informal and somewhat freeform.

Although there are many scenarios in which this extra metadata may be picked up and interpreted by a user-agent, the idea is that they are not _specific_ to any particular user-agent and can be safely ignored (in that sense they are similar to properties, although the difference is that they are still data). The notion of a twitter-friendly URL for a piece of content can be very handy for applications, but it's not specific to any one renderer.

Regardless, I do completely understand your point here. Let me propose three options as I see them:
1. **Freeform, top level**: The model I was explaining above. Simplest for authors to implement (no prefixes or extra objects). However, there will be more heterogeneity between `book.json` files in the wild, potentially leading to confusion.
2. **Prefixed, top level**: Similar to above, but non-spec properties receive a prefix, perhaps something like `data-propname` used in HTML5 or CSS vendor-style prefixes like `-baker-propname`
3. **Dedicated object**: All author custom, user-agent specific, or otherwise non-standard properties go into one or more dedicated objects for storage. Likely the most futureproof, possibly cleaner in terms of implementation.

Did I miss any option? As is obvious, I have a preference toward 1 and 2. However, this all ties in somewhat into the final item.

**Properties vs. x-baker**

> So orientation and pinchToZoom are in this perspective properties of the book and the rendering engine should read these and adapt accordingly. Instead, the extensions are add-ons, out of the standard, so a book might not use them at all. While not specifying the orientation is a problem, because all the tablets out there can manage that, ignoring "background" is just something that affect the specific "Baker" rendering engine, and another reader, for example Treesaver, can safely ignore them

This is the tricky part -- one man's universal, inherent property is another's optional add-on :) For a tablet-only (iPad-only?) project such as Baker, `pinchToZoom` and `orientation` are seen as first-class aspects of a publication. However, these aspects aren't universal once you move to laptops, TVs, etc.

Again, I can see both sides of the argument, I just have a preference toward the flatter model (prefix or not), for these reasons:
- Fewer distinct objects means less typing and structure
- Easier to document/teach authors to create the file, since there are fewer buckets where properties can go (i.e. don't have to think whether `background` goes in `properties` or `x-baker`?)
- Has parallels to HTML/CSS

But of course, I know there are disadvantages as well -- which is why this is a preference :)

**Bonus: Hierarchy**

Our scenarios require hierarchy for the content of the publication, for example a newspaper has sections for Fashion, Sports, etc. I see that you have some requests for supporting multiple HPubs, so perhaps there will be some parallels for you as well.

@BramStein and I have talked about a few options, but one that is interesting is to simply allow full nesting within a contents item -- meaning each article could have roughly the same properties as the root publication, including a `contents` array. Obviously though, there are some tricky aspects here -- but something worth talking about at least.

If hierarchy isn't something Baker is planning on doing any time soon, then we'll need to have our own method, based on whichever extensibility option we choose :)

**Comment by fortes** [comment]
On a different note, I'm curious why you chose for `creator` to be a single string instead of an array. Multiple authors is something that happens frequently, and unless you prescribe a strict standard (comma after every name, etc), then extracting authors could get tricky. Perhaps this is another case where one can either use a plain string, or choose to use an array.

Package.json uses an array w/ hashes for various properties in this case (to hold web links, twitter, etc). That may be overkill here, depending on how you see those fields being used.

**Comment by folletto** [comment]
Thanks for the follow up. :)
1. _Ok_ for "contents". :)
2. _Ok_ for page hashes in contents. Ok for url being the only required one.
3. Metadata: I think we have to split this in two. Pages metadata and Book metadata.
4. **Page metadata**: I can agree that anything could be added here. I can't see any harm in adding more stuff than the specified one, even if they become vendor specific, it's quite impossible to add anything else but metadata here and being already nested two times inside both contents and page, makes it quite restrictive already.
5. **Book metadata**: This for me is trickier, and again it's splitted in different topics. I'll take an example to show you why. 
   1. _On explicit extension naming (x- or similar)_: Let's say someone adds a "license" attribute, that looks like as native, and it's a string. Then, someone else adds another "license" attribute, but he makes it an hash with a string for the readable name, a url for the full license and a MD5 hash to identify it in their system. What happens then? A book shared between the two will show random data or behave strangely. And, after that, we can't anymore standardize it because the "license" parameter has already been taken and we will break any past book if we norm that after the problem happened. 
   2. _On grouping_: I think this allows more flexiblity: as Baker I have to take care only of the x-baker block, and you as Treesaver can ignore it safely, instead of iterating through all the elements. Also, using a block like that optimizes typing: instead of typing -baker-background, -baker-verticalBounce, -baker-indexHeigh, -baker-mediaAutoPlay we'll type x-baker once and it's done. And "background" in Baker is a very specific thing, that is going probably to exist in Baker only, until we remove it because it's just a workaround. So, it would be an even bigger problem if ALL the authors everywhere needs to take care of the background property. Plus in this way at root level we have just metadata and structure, no strange things, it's very clean.
   3. _On proper metadata_: If you really think that Hpub is missing a metadata, you can start adding it in your client x- section, but then you can come here, discuss, and we'll add it if there's enough consensus in the community.
   4. _My stance_: of this list, I think that 1 is really important, 2 is just my preference, but I see your point here, and 3 is a very important part to progress the standard and to aggregate the community. So, if we want to meet in the middle ground, I think that your option (2) above is the one I'd choose: "Prefixed, top level".
6. **Author/Creator**: _Ok_, I agree also here on your suggestion of making them arrays, and I agree in making it not overly complex: just names.
7. **What is a property**: Given your opinion, I agree on how "pinchToZoom" looks device-dependent. It should be probably named "zoomable" or something like that. A publication supporting "zoomable" and "orientation" to me needs to behave that way in any device. A browser is zoomable, for example even if it doesn't support pinch, and disallowing or allowing so is a choice in the hands of the author. Orientation is kinda the same: a device needs to know if it can turn the page sideways when needed (through a gyroscope, but why not, even through a specific UI element). So, these are strictly related to the book. Instead, things like "x-baker/indexHeight" will be used only in Baker and control a _specific_ way Baker handles the index overlay... until we are able to automate that. :) But I think that this point is solved for you if we choose (2) "Prefixed, top level" right? :)
8. **Nested contents**: It's not for Baker, but I can see why you raise that point. We can reserve right now the word "contents" inside a page, it doesn't do harm, but we might avoid to spec that out until some more discussion at a later time. Or do you think it's urgent? :)

**Comment by fortes** [comment]
I think we're getting close here, but I realize that I should clarify something from Treesaver's part (so you can better understand my viewpoint). There are very few things in the JSON that Treesaver will natively use. Specifically, many scenarios will likely only use the `contents` array, and just as a way to string articles together in a specific order. Treesaver-specific settings may get a decent amount of use as well, although that hasn't be exposed quite yet so time will tell.

All other metadata, be it at the top-level or on a specific article is used solely at the discretion of the developer using the Treesaver framework. Developers use HTML templates to define the UI shown by Treesaver. Within this HTML, developers can bind to variables such as the current page number & page count. Additionally, they can supply their own metadata on the article/publication level for use within the HTML. These might be used to create a running title, or provide a shortened link for twitter, or even a sidebar of navigation with titles and thumbnails for each article --- all via binding to user-provided metadata, there is no native Treesaver notion of Table of Contents UI.

So that means the only properties we absolutely care about, and will base behavior around are the `contents` array and however we decide to do properties/settings. By default, Treesaver has no use for `title`, `author`, etc -- it will only be accessed if the user explicitly builds UI that binds to that variable name.

So, my general priorities here are:
1. Make sure our scenarios are possible
2. Developer ease of use / simple documentation
3. Easier implementation

OK, now that I've finished that long-winded story -- here's my take on what we've been discussing:

**Book/Page Metadata**

Our scenarios are covered by being able to add metadata, and the implementation shouldn't be an issue -- so the only thing to think about here is developer usability / ease of documentation. 

In order to make the concepts easier, I think whatever we do for the article/page level should be consistent with the publication level. 

**Naming & Grouping**

Your point about standardization and future-proofing is a good one, so I think we both agree that prefixing of some sort is the way to go here. I personally think it's easier to explain/document letting custom fields be peers of other attributes that differ in that they have special prefix. If you feel strongly about boxing this into a custom bucket, let me know -- but I think it's simpler not to, even if you do pay the price of repetition. (Implementation-wise, it's trivial to ignore any key that doesn't have your prefix).

So, to update your example above, here's what I think we're converging to (let me know if I've missed something or made a bad assumption):

``` js
{
  // Standard properties. No UA requirement to do anything with them though
  "title": "The Study in Scarlet",
  "author": "Arthur Conan Doyle",
  "creator": ["D. Casali", "M. Colombo", "A. Morandi", "F. Fortes"],
  "date: "2011-08-23",
  "url": "book://bakerframework.com/books/arthurconandoyle-thestudyinscarlet",

  // I prefer settings, but I'm fine with properties if you'd prefer that :)
  "settings": {
    "orientation": "both",
    "pinchToZoom": false,

    // Custom user-agent properties are prefixed. Need to chose between either be via x-ua-
    "x-baker-background": "#000000",
    "x-baker-verticalBounce": true,
    // Or CSS-style -ua-
    "-baker-indexHeight": 150,
    "-baker-mediaAutoPlay": true

    // Other UAs may have their own settings
    "-treesaver-loopArticles": "true"
  },

  "contents": [
    "Book Cover.html",
    "Book Index.html",
    {
        // Only URL is required
        "url": "Front-Article.html",
        // Contents is reserved?

        // All UA-specific properties need prefixing
        "-treesaver-requires": "touch"

        // Have generic prefix for non-UA custom metadata?
        "custom-title": "The Wonders of Hpub",
        // Follow HTML5's lead?
        "data-author": "The community"
        // Go for brevity and make prefix a single -?
        "-shortlink": "http://t.co/foobar"
    },
    "Part1-01.html",
    "Part1-02.html",
    "Part1-03.html",
    "Part1-04.html",
    "Part1-05.html",
    "Part1-06.html",
    "Part1-07.html",
    "Part2-01.html",
    "Part2-02.html",
    "Part2-03.html",
    "Part2-04.html",
    "Part2-05.html",
    "Part2-06.html",
    "Part2-07.html",
    "Colophon.html",
  ],

  // Top-level custom metadata
  "custom-issueNumber": 4,
  "data-twitter": "@BakerFramework"
}
```

Does that line up with your thinking? Or did I misunderstand?

**Comment by fortes** [comment]
Less important, since Treesaver won't do anything with this natively -- I thought about the author / creator fields a bit more, and I think it may be worth providing support for "string" or {object}, just like with the `contents` array -- URL links, twitter handles, etc can be very useful. But I've got no horse in that race, so up to you :)

**Comment by bramstein** [comment]
Minor detail perhaps, but wouldn't it be easier to document if both top level and per article/document settings were specified in the same way (i.e. through the `settings` property?)

**Comment by folletto** [comment]
Yes. I agree @bramstein.

@Fortes, could you check my update to the top example (first message of this page)? Does it work for you? Is there some mismatch? :)

**Comment by fortes** [comment]
The example above works fine for Treesaver. Is the Baker team comfortable w/ it as well?

Related, I realized that this is flexible enough that UAs can still choose to use a settings object if they prefer:

``` js
  "-ua-settings": {
    "setting": "value"
  }
```

I think we've got something fairly simple here and easy to explain. As we get implementations out there, we'll get feedback on what's confusing and what works. Then we can change as we see fit. I'm open to most anything, as long as it doesn't make our scenarios difficult or impossible. If it turns out that everyone prefers more nesting, etc -- then who am I to disagree? :)

**Comment by folletto** [comment]
Yes, we'll double check with the others when they are back, but I think that's the version that makes everyone happy. :)
I also agree with you for the future evolution, that's exactly what I'm hoping for. :)

(btw, nice spotting of that kind of behaviour possibility) :D

**Comment by bramstein** [comment]
There's another use-case that is mostly Treesaver specific, but perhaps of interest to Baker as well. As @fortes mentioned, most of our meta-data is completely developer specific, apart from `contents` and `url`. A developer can enter any properties as meta-data and use those properties in templates.

I recently had a case where it was useful to have the framework set some "status" properties on the meta-data, such as `active` or `visited` on items in the `contents` array. This would allow the developer to implement custom styling in their templates for currently active or previously visited articles, sections or chapters. There is however a risk of name clashes between properties created by the framework and those created by the developer. Should such properties also be prefixed, or should they go into a "reserved" property (for example, one called "status".)

**Comment by fortes** [comment]
@bramstein I think we prefix any built-in/native behavior until we're ready to propose them as part of the HPub spec.

**Comment by folletto** [comment]
@bramstein Uhm given what I understand from your explanation it seems a behaviour that derives from some development choices you are giving your developers, so I think it could be something that Treesaver is free to manage inside its own custom parameters. :)

And then, as @fortes was saying, if something very solid and of common interests emerges between all your users, we can just discuss in the following standard, mostly like CSS. :)

**Comment by fortes** [comment]
You bring up a good point, @folletto, should there be a `hpub-version` property?

**Comment by folletto** [comment]
Wow that would have been a big oversight. ^^

What do you think?

``` js
  "hpub": 1.0
```

Would be ok? :)

**Comment by fortes** [comment]
Works for me.

**Comment by Simbul** [comment]
Manifest should now be properly implemented in master, since commit 8fc8d53f956bec2977c13652606ad1620c039a62.
Please test it and let us know if it's working! :)
